File: lib/turnstile.js
const { delay } = require("./util");
const { log } = require("./logging");

// Detects a Cloudflare Turnstile challenge iframe on the page.
async function isTurnstilePresent(page) {
  try {
    return await page.evaluate(() => {
      const iframes = [...document.querySelectorAll("iframe")];
      return iframes.some(f => (f.src || "").includes("challenges.cloudflare.com"));
    });
  } catch (_) {
    return false;
  }
}

// Polls for a Turnstile challenge over several seconds instead of checking
// once after a fixed delay — Cloudflare's injection timing varies, and a
// single early check can miss a challenge that renders a bit slower.
//
// Logs each iteration's outcome (with elapsed time) so a poll that exits
// earlier than its nominal timeoutMs is visible in the logs instead of
// showing up as an unexplained gap between two timestamps — previously
// indistinguishable from "genuinely no Turnstile in 6s" versus "the loop
// exited early for some other reason" (e.g. an evaluate() error from the
// page navigating mid-check), which made a real occurrence hard to
// diagnose after the fact. Uses its own try/catch around evaluate()
// directly (rather than calling isTurnstilePresent) so it can log the
// distinction between "not found" and "check errored" per iteration.
async function pollForTurnstile(page, timeoutMs = 6000, intervalMs = 750, shortT = "?") {
  const start = Date.now();
  let iteration = 0;
  while (Date.now() - start < timeoutMs) {
    iteration++;
    const elapsed = Date.now() - start;
    try {
      const found = await page.evaluate(() => {
        const iframes = [...document.querySelectorAll("iframe")];
        return iframes.some(f => (f.src || "").includes("challenges.cloudflare.com"));
      });
      if (found) {
        log(`[turnstile:${shortT}] poll #${iteration} at ${elapsed}ms: Turnstile iframe found`);
        return true;
      }
      log(`[turnstile:${shortT}] poll #${iteration} at ${elapsed}ms: not found yet`);
    } catch (err) {
      log(`[turnstile:${shortT}] poll #${iteration} at ${elapsed}ms: evaluate() error (likely mid-navigation) — ${err.message}`);
    }
    await delay(intervalMs);
  }
  log(`[turnstile:${shortT}] poll window closed after ${iteration} iteration(s), ${Date.now() - start}ms elapsed — no Turnstile found`);
  return false;
}

// Extracts the Turnstile sitekey from the page. Checks the common places it
// appears: a [data-sitekey] element, or embedded in the challenge iframe's src.
async function getTurnstileSitekey(page) {
  try {
    return await page.evaluate(() => {
      const el = document.querySelector("[data-sitekey]");
      if (el) return el.getAttribute("data-sitekey");

      const iframe = [...document.querySelectorAll("iframe")]
        .find(f => (f.src || "").includes("challenges.cloudflare.com"));
      if (iframe) {
        try {
          const url = new URL(iframe.src);
          // sitekey commonly appears as a path segment after /turnstile/if/ove/
          const match = url.pathname.match(/\/([0-9a-zA-Z_-]{20,})\//);
          if (match) return match[1];
        } catch (_) {}
      }
      return null;
    });
  } catch (_) {
    return null;
  }
}

// Injects a solved Turnstile token into the page's hidden response field,
// matching what a real solve would populate.
async function injectTurnstileToken(page, tokenValue) {
  return page.evaluate((val) => {
    const candidates = [
      'input[name="cf-turnstile-response"]',
      'textarea[name="cf-turnstile-response"]',
      'input[name="g-recaptcha-response"]', // some Turnstile embeds reuse this name
    ];
    for (const sel of candidates) {
      const el = document.querySelector(sel);
      if (el) {
        el.value = val;
        el.dispatchEvent(new Event("input", { bubbles: true }));
        el.dispatchEvent(new Event("change", { bubbles: true }));
        return true;
      }
    }
    return false;
  }, tokenValue);
}

// Submits a Turnstile solve task to the self-hosted icemellow-me solver
// (2captcha-compatible API) and polls until solved, failed, or timed out.
// Returns the token string, or null if solving failed/unavailable.
// Reads env vars fresh on each call (rather than once at module load) so
// credentials can be rotated via Railway without requiring a redeploy.
async function solveTurnstile(sitekey, pageUrl, shortT) {
  const SOLVER_URL = (process.env.TURNSTILE_SOLVER_URL || "").replace(/\/$/, "");
  const SOLVER_KEY = process.env.TURNSTILE_SOLVER_KEY || "";

  if (!SOLVER_URL || !SOLVER_KEY) {
    log(`[turnstile-solver:${shortT}] TURNSTILE_SOLVER_URL/KEY not configured, skipping solve attempt`);
    return null;
  }
  if (!sitekey) {
    log(`[turnstile-solver:${shortT}] Could not extract sitekey, skipping solve attempt`);
    return null;
  }

  try {
    const submitUrl = `${SOLVER_URL}/in.php?key=${encodeURIComponent(SOLVER_KEY)}&method=turnstile&sitekey=${encodeURIComponent(sitekey)}&pageurl=${encodeURIComponent(pageUrl)}&json=1`;
    const submitRes = await fetch(submitUrl, { signal: AbortSignal.timeout(15000) });
    const submitJson = await submitRes.json();
    if (submitJson.status !== 1) {
      log(`[turnstile-solver:${shortT}] Submit failed: ${JSON.stringify(submitJson)}`);
      return null;
    }
    const taskId = submitJson.request;
    log(`[turnstile-solver:${shortT}] Task submitted (${taskId}), polling...`);

    const POLL_TIMEOUT_MS = 60_000;
    const POLL_INTERVAL_MS = 3000;
    const start = Date.now();

    while (Date.now() - start < POLL_TIMEOUT_MS) {
      await delay(POLL_INTERVAL_MS);
      const pollUrl = `${SOLVER_URL}/res.php?key=${encodeURIComponent(SOLVER_KEY)}&id=${encodeURIComponent(taskId)}&json=1`;
      const pollRes = await fetch(pollUrl, { signal: AbortSignal.timeout(10000) });
      const pollJson = await pollRes.json();

      if (pollJson.status === 1) {
        log(`[turnstile-solver:${shortT}] Solved after ${Math.round((Date.now() - start) / 1000)}s`);
        return pollJson.request;
      }
      if (pollJson.request && pollJson.request !== "CAPCHA_NOT_READY") {
        log(`[turnstile-solver:${shortT}] Solve failed: ${pollJson.request}`);
        return null;
      }
    }
    log(`[turnstile-solver:${shortT}] Solve timed out after ${POLL_TIMEOUT_MS / 1000}s`);
    return null;
  } catch (err) {
    log(`[turnstile-solver:${shortT}] Solver request error: ${err.message}`);
    return null;
  }
}

// Combines sitekey extraction, calling the solver API, and injecting the
// resulting token into the page. Returns true if a token was successfully
// obtained and injected, false otherwise (caller should fall back to skip).
async function trySolveTurnstileOnPage(page, shortT) {
  const sitekey = await getTurnstileSitekey(page);
  const pageUrl = page.url();
  const token = await solveTurnstile(sitekey, pageUrl, shortT);
  if (!token) return false;

  const injected = await injectTurnstileToken(page, token);
  if (!injected) {
    log(`[turnstile-solver:${shortT}] Got token but found no field to inject it into`);
    return false;
  }
  return true;
}

module.exports = {
  isTurnstilePresent,
  pollForTurnstile,
  getTurnstileSitekey,
  injectTurnstileToken,
  solveTurnstile,
  trySolveTurnstileOnPage,
};

File: lib/turnstile.py
"""Port of lib/turnstile.js.

The external-solver-API flow (sitekey extraction -> HTTP solve -> token
injection) is replaced entirely by playwright_captcha's ClickSolver, which
clicks the real Turnstile checkbox in-page. This removes the need for
TURNSTILE_SOLVER_URL/KEY, a separate solver service, and sitekey capture
hooks in browser.py.
"""
import time

from .util import delay
from .logging import log, log_verbose
from .screenshot import save_debug_html


async def is_turnstile_present(page, short_t="?") -> bool:
    try:
        result = await page.evaluate("""
            () => {
                const hasHostDiv = !!document.querySelector('div[id^="cf-turnstile"], div.cf-turnstile, [data-sitekey]');
                const iframes = [...document.querySelectorAll("iframe")];
                const srcs = iframes.map(f => f.src || "(no src)");
                const hasMatchingIframeSrc = iframes.some(f => (f.src || "").includes("challenges.cloudflare.com"));
                let foundVia = null;
                if (hasHostDiv) foundVia = "host-div";
                else if (hasMatchingIframeSrc) foundVia = "iframe-src";
                return { found: hasHostDiv || hasMatchingIframeSrc, foundVia, allSrcs: srcs };
            }
        """)
        found = result["found"]
        found_via = result["foundVia"]
        all_srcs = result["allSrcs"]

        if found:
            log_verbose(f"[turnstile-diag:{short_t}] Turnstile detected via {found_via}")
        elif all_srcs:
            log_verbose(f"[turnstile-diag:{short_t}] no Turnstile host-div or matching iframe; {len(all_srcs)} iframe(s) present: {all_srcs}")
        else:
            log_verbose(f"[turnstile-diag:{short_t}] no Turnstile host-div and no iframes present at all on this check")
        return found
    except Exception as err:
        log_verbose(f"[turnstile-diag:{short_t}] isTurnstilePresent evaluate() error: {err}")
        return False


async def is_cloudflare_interstitial(page) -> bool:
    try:
        return await page.evaluate("""
            () => {
                const bodyText = document.body.innerText || "";
                const patterns = [
                    /performing security verification/i,
                    /checking your browser/i,
                    /attention required/i,
                    /just a moment/i,
                ];
                if (patterns.some(p => p.test(bodyText))) return true;
                const scripts = [...document.querySelectorAll("script[src], iframe[src]")];
                return scripts.some(el => /\\/cdn-cgi\\/|\\/challenge-platform\\//.test(el.src || ""));
            }
        """)
    except Exception:
        return False


async def get_post_click_page_state(page, short_t="?") -> str:
    """Returns one of: "success" | "already" | "interstitial" | "captcha" | "unknown"."""
    try:
        if await is_cloudflare_interstitial(page):
            return "interstitial"

        dom_state = await page.evaluate("""
            () => {
                const bodyText = document.body.innerText || "";
                if ([...document.querySelectorAll("h2")].some(h => /thanks for voting/i.test(h.textContent))) {
                    return "success";
                }
                if (/you have already voted/i.test(bodyText) || /you can vote again/i.test(bodyText)) {
                    return "already";
                }
                return null;
            }
        """)
        if dom_state:
            return dom_state

        if await is_turnstile_present(page, short_t):
            return "captcha"

        return "unknown"
    except Exception:
        return "unknown"


async def poll_post_click_state(page, timeout_ms=20000, interval_ms=750, short_t="?") -> str:
    start = time.monotonic()
    iteration = 0
    while (time.monotonic() - start) * 1000 < timeout_ms:
        iteration += 1
        elapsed = (time.monotonic() - start) * 1000
        state = await get_post_click_page_state(page, short_t)
        if state != "unknown":
            log_verbose(f"[vote-state:{short_t}] poll #{iteration} at {elapsed:.0f}ms: {state}")
            return state
        log_verbose(f"[vote-state:{short_t}] poll #{iteration} at {elapsed:.0f}ms: unknown, still waiting")
        await delay(interval_ms)
    log(f"[vote-state:{short_t}] Post-click state poll timed out after {(time.monotonic() - start) * 1000:.0f}ms with no resolved state")
    return "unknown"


async def poll_for_turnstile(page, timeout_ms=6000, interval_ms=750, short_t="?") -> bool:
    start = time.monotonic()
    iteration = 0
    while (time.monotonic() - start) * 1000 < timeout_ms:
        iteration += 1
        elapsed = (time.monotonic() - start) * 1000
        try:
            found = await page.evaluate("""
                () => {
                    const hasHostDiv = !!document.querySelector('div[id^="cf-turnstile"], div.cf-turnstile, [data-sitekey]');
                    const iframes = [...document.querySelectorAll("iframe")];
                    const hasMatchingIframeSrc = iframes.some(f => (f.src || "").includes("challenges.cloudflare.com"));
                    return hasHostDiv || hasMatchingIframeSrc;
                }
            """)
            if found:
                log_verbose(f"[turnstile:{short_t}] poll #{iteration} at {elapsed:.0f}ms: Turnstile iframe found")
                return True
            log_verbose(f"[turnstile:{short_t}] poll #{iteration} at {elapsed:.0f}ms: not found yet")
        except Exception as err:
            log_verbose(f"[turnstile:{short_t}] poll #{iteration} at {elapsed:.0f}ms: evaluate() error (likely mid-navigation) — {err}")
        await delay(interval_ms)
    log_verbose(f"[turnstile:{short_t}] poll window closed after {iteration} iteration(s), {(time.monotonic() - start) * 1000:.0f}ms elapsed — no Turnstile found")
    return False


async def try_solve_turnstile_on_page(page, short_t="?", bot_id="?") -> bool:
    """Solves an in-page Turnstile challenge using playwright_captcha's
    ClickSolver against the voter's own live page/context — clicks the real
    checkbox element (found via shadow-root/light-DOM traversal), rather
    than the old sitekey-extraction + external-API + token-injection flow.

    Cloudflare can inject the actual interactive checkbox into the iframe
    some time after the outer widget container first appears in the DOM —
    the container existing (what is_turnstile_present checks) doesn't
    guarantee the checkbox inside is ready yet. ClickSolver's own internal
    attempt budget can run out during that gap. Rather than trust a single
    ClickSolver call, this re-attempts it in an outer loop as long as our
    own reliable detection still shows the widget present, so a
    late-appearing checkbox gets caught on a subsequent pass instead of
    the whole solve being abandoned after one internal timeout."""
    try:
        from playwright_captcha import CaptchaType, ClickSolver, FrameworkType
    except ImportError as e:
        log(f"[turnstile-solver:{short_t}] playwright_captcha not available: {e}")
        return False

    OUTER_MAX_ATTEMPTS = 3
    OUTER_RETRY_DELAY_S = 5

    last_err = None
    for outer_attempt in range(1, OUTER_MAX_ATTEMPTS + 1):
        if not await is_turnstile_present(page, short_t):
            log(f"[turnstile-solver:{short_t}] Turnstile no longer present for bot {bot_id} (resolved on its own or page moved on) — treating as solved")
            return True

        try:
            async with ClickSolver(framework=FrameworkType.CAMOUFOX, page=page) as solver:
                await solver.solve_captcha(
                    captcha_container=page,
                    captcha_type=CaptchaType.CLOUDFLARE_TURNSTILE,
                )
            log(f"[turnstile-solver:{short_t}] ClickSolver solved Turnstile for bot {bot_id} (outer attempt {outer_attempt})")
            return True
        except Exception as e:
            last_err = e
            if outer_attempt < OUTER_MAX_ATTEMPTS:
                log(f"[turnstile-solver:{short_t}] ClickSolver attempt {outer_attempt}/{OUTER_MAX_ATTEMPTS} failed for bot {bot_id} ({e}) — checkbox may not have been ready yet, retrying in {OUTER_RETRY_DELAY_S}s...")
                await delay(OUTER_RETRY_DELAY_S * 1000)

    log(f"[turnstile-solver:{short_t}] ClickSolver could not solve Turnstile for bot {bot_id} after {OUTER_MAX_ATTEMPTS} outer attempts: {last_err}")
    await save_debug_html(page, short_t, bot_id, "turnstile-clicksolver-failed")
    return False


async def try_solve_interstitial_on_page(page, short_t="?", bot_id="?") -> bool:
    """Attempts to resolve a full-page Cloudflare interstitial using
    playwright_captcha's ClickSolver (CLOUDFLARE_INTERSTITIAL) — an active
    click attempt, tried before falling back to passively waiting for
    Cloudflare's own redirect. Interstitials are frequently non-interactive
    (a background JS/proof-of-work check with no checkbox at all), so this
    is expected to no-op harmlessly in that case; ClickSolver raises if it
    can't find a clickable challenge, which the caller should treat the same
    as "still waiting", not as a hard failure."""
    try:
        from playwright_captcha import CaptchaType, ClickSolver, FrameworkType
    except ImportError as e:
        log(f"[interstitial-solver:{short_t}] playwright_captcha not available: {e}")
        return False

    try:
        async with ClickSolver(framework=FrameworkType.CAMOUFOX, page=page) as solver:
            await solver.solve_captcha(
                captcha_container=page,
                captcha_type=CaptchaType.CLOUDFLARE_INTERSTITIAL,
            )
        log(f"[interstitial-solver:{short_t}] ClickSolver resolved interstitial for bot {bot_id}")
        return True
    except Exception as e:
        log_verbose(f"[interstitial-solver:{short_t}] ClickSolver did not resolve interstitial for bot {bot_id} (may be non-interactive): {e}")
        return False
